src/cocaine-app/infrastructure.py

```diff
@@ -167,158 +166,244 @@ def _sync_state(self):
                 self._sync_state
             )
 
-    @classmethod
-    def __node_state_type(cls, node_state):
-        return node_state.get(
-            'type',
-            (
-                node_state.get('manual', False) and cls.HISTORY_RECORD_MANUAL or
-                cls.HISTORY_RECORD_AUTOMATIC
+    def __do_sync_state(self):
+        """
+        Apply all new non-automatic history records
+
+        To sync group state among all mastermind workers we apply the following strategy:
+
+            any worker can add a new state record (couple or node backends set) and
+            mark it with a non-automatic type (manual, job, etc.). Such state changes
+            are found and applied by all other workers periodically.
+
+        This method implements the described strategy by performing search and applying
+        records that are found.
+        """
+        new_ts = int(time.time())
+
+        for group_history in self.group_history_finder.search_by_history_record(
+            type=(GroupStateRecord.HISTORY_RECORD_MANUAL, GroupStateRecord.HISTORY_RECORD_JOB),
+            start_ts=self._sync_ts,
+            finish_ts=new_ts
+        ):
+            logger.debug('Found updated group history for group {}'.format(group_history.group_id))
+            if group_history.group_id not in storage.groups:
+                continue
+            group = storage.groups[group_history.group_id]
+            for node_backends_set in group_history.nodes:
+                if not self._sync_ts <= node_backends_set.timestamp < new_ts:
+                    continue
+                if node_backends_set.type not in (GroupStateRecord.HISTORY_RECORD_MANUAL,
+                                                  GroupStateRecord.HISTORY_RECORD_JOB):
+                    continue
+
+                for nb in group.node_backends:
+                    for nb_record in node_backends_set.set:
+                        if (
+                            nb_record.hostname == nb.hostname and
+                            nb_record.port == nb.port and
+                            nb_record.family == nb.family and
+                            nb_record.backend_id == nb.backend_id and
+                            nb_record.path == nb.base_path
+                        ):
```

Comment:
Is it `GroupNodeBackendRecord`? You have `def __eq__` for `GroupNodeBackendRecord`.